src/kemal/middleware/filters.cr

```diff
@@ -35,18 +47,25 @@ module Kemal::Middleware
     def initialize(&@block : HTTP::Server::Context -> _)
     end
   end
+
 end
 
 def add_filters
-  Kemal.config.add_handler Kemal::Middleware::Filter.new
+  unless filter = Kemal.config.handlers.any? { |handler| handler.is_a? Kemal::Middleware::Filter }
+    filter = Kemal::Middleware::Filter.new
+    Kemal.config.add_handler filter
+  end
+  filter
 end
 
 def before(path = "*", options = {} of Symbol => String, &block : HTTP::Server::Context -> _)
-  filter = Kemal.config.handlers.first as Kemal::Middleware::Filter
+  filter = (Kemal.config.handlers.find { |handler| handler.is_a? Kemal::Middleware::Filter } || add_filters) as Kemal::Middleware::Filter
+  raise Kemal::Middleware::Filter::BeforeFilterAlreadyDefinedException.new("A before-filter already is already defined for path: '#{path}'.") if filter.filter_for_path_type_defined?(path, :before)
   filter.add :before, path, options, &block
 end
 
 def after(path = "*", options = {} of Symbol => String, &block : HTTP::Server::Context -> _)
-  filter = Kemal.config.handlers.first as Kemal::Middleware::Filter
+  filter = (Kemal.config.handlers.find { |handler| handler.is_a? Kemal::Middleware::Filter } || add_filters) as Kemal::Middleware::Filter
+  raise Kemal::Middleware::Filter::AfterFilterAlreadyDefinedException.new("An after-filter already is already defined for path: '#{path}'.") if filter.filter_for_path_type_defined?(path, :after)
```

Comment:
How about having the exception message in exception class itself instead of initialization?